test/ts/unit/collateralized_contract.ts

```diff
@@ -0,0 +1,508 @@
+            )).to.eventually.be.rejectedWith(REVERT_ERROR);
+        });
+
+      });
+
+    });
+
+    describe("#returnCollateral", () => {
+
+      describe("#invariants", () => {
+
+        it("should throw if no collateral is mapped to the agreement id", async () => {
+            const ID_FOR_NON_EXISTENT_AGREEMENT = web3.sha3("this agreement does not exist.");
+
+            await expect(collateralContract.returnCollateral.sendTransactionAsync(
+                ID_FOR_NON_EXISTENT_AGREEMENT
+            )).to.eventually.be.rejectedWith(REVERT_ERROR);
+        });
+
+        it("should throw if the lockup period is still in effect", async () => {
+            const ID_FOR_ACTIVE_AGREEMENT = web3.sha3("this agreement is still in effect.");
+
+            // Collateralize an agreement with a lock up period in the future.
+            await collateralContract.setDummyCollateral.sendTransactionAsync(
+                ID_FOR_ACTIVE_AGREEMENT,
+                COLLATERALIZER,
+                mockToken.address,
+                COLLATERAL_AMOUNT,
+                new BigNumber(moment().add(2, 'years').unix()), // lockup period is still in effect.
+                false
+            );
+
+            await expect(collateralContract.returnCollateral.sendTransactionAsync(
+                ID_FOR_ACTIVE_AGREEMENT
+            )).to.eventually.be.rejectedWith(REVERT_ERROR);
+        });
+
+        it("should throw if the collateral has already been withdrawn", async () => {
+            const ID_FOR_RESOLVED_AGREEMENT = web3.sha3("this agreement has been resolved.");
+
+            // Collateralize an agreement with a lock up period in the future.
+            await collateralContract.setDummyCollateral.sendTransactionAsync(
+                ID_FOR_RESOLVED_AGREEMENT,
+                COLLATERALIZER,
+                mockToken.address,
+                COLLATERAL_AMOUNT,
+                new BigNumber(moment().subtract(2, 'years').unix()),
+                true // collateral marked as withdrawn.
+            );
+
+            await expect(collateralContract.returnCollateral.sendTransactionAsync(
+                ID_FOR_RESOLVED_AGREEMENT
+            )).to.eventually.be.rejectedWith(REVERT_ERROR);
+        });
+
+        it("should throw if the agreement is in default and the lockup period has expired", async () => {
+            const DEFAULTED_AGREEMENT_ID = web3.sha3("this agreement was defaulted upon.");
+
+            await collateralContract.setDummyCollateral.sendTransactionAsync(
+                DEFAULTED_AGREEMENT_ID,
+                COLLATERALIZER,
+                mockToken.address,
+                COLLATERAL_AMOUNT,
+                new BigNumber(moment().subtract(1, 'month').unix()), // lockup period has expired.
+                false
+            );
+
+            await collateralContract.setDummyExpectedRepaymentValue.sendTransactionAsync(
+                DEFAULTED_AGREEMENT_ID,
+                COLLATERAL_AMOUNT
```

Comment:
This test should explicitly set the value repaid to be less than the expectedRepaymentValue -- right now, it is implicitly defaulting be cause `getValueRepaidToDate` hasn't been mocked for the `DEFAULTED_AGREEMENT_ID` yet

Reply:
Agreed. Good call.